### PR TITLE
Touchup "Quickstart Guide: Compose and Django"

### PR DESCRIPTION
**Pull request by charleswhchan**
Signed-off-by: Charles Chan <charleswhchan@users.noreply.github.com>

### CODE REVIEW

File: docs/django.md [hunk 1]
```diff
@@ -12,18 +12,16 @@ weight=4
...
 ### Define the project
 
-Start by setting up the three files you'll need to build the app. First, since
+For this project, you need to define three files. First, since
 your app is going to run inside a Docker container containing all of its
-dependencies, you'll need to define exactly what needs to be included in the
-container. This is done using a file called `Dockerfile`. To begin with, the
-Dockerfile consists of:
+dependencies, you'll need a `Dockerfile` to define exactly what needs to be
+included in the image. To begin with, create a `Dockerfile` consists of:
```
Comment (aanand) [inline docs/django.md]: "image" is arguably a bit more technically accurate, but for new users, "container" might be less confusing (and it's not wrong).

**Comment by aanand** [comment]
ping @moxiegirl

File: docs/django.md [hunk 2]
```diff
@@ -12,18 +12,16 @@ weight=4
 
 ## Quickstart Guide: Compose and Django
```
Comment (moxiegirl) [inline docs/django.md]: @charleswhchan Thanks for the changes Charles, are you willing to go a bit further? These instructions are very hard to follow because they don't use steps which can help readers follow the flow better.  Would you be willing to add them ? There are other structural issues like it stars with a head ## instead of a # ....

Having you do this means we get an extra tech review on the steps.

--> https://gist.github.com/moxiegirl/e4b95e07a7f811db40b5





# Quickstart Guide: Compose and Django

This quick-start guide demonstrates how to use Compose to set up and run a
simple Django/PostgreSQL app. Before starting, you'll need to have
[Compose installed](install.md).

## Define the project components

For this project, you need to create a Dockerfile, a Python dependencies file,
and a `docker-compose.yml` file.

1. Create an empty project directory.

    You can name the directory something easy for your to remember. This directory is the context for your application image. The directory should only contain resources to build that image.

2. Create a new file called `Dockerfile` in your project directory.

    The Dockerfile defines an application's image content via one or more build
    commands that configure that image. Once built, you can run the image in a
    container.  For more information on `Dockerfiles`, see the [Docker user
    guide](https://docs.docker.com/userguide/dockerimages/#building-an-image-from-a-dockerfile)
    and the [Dockerfile reference](http://docs.docker.com/reference/builder/).

3. Add the following content to the `Dockerfile`.

        FROM python:2.7
        ENV PYTHONUNBUFFERED 1
        RUN mkdir /code
        WORKDIR /code
        ADD requirements.txt /code/
        RUN pip install -r requirements.txt
        ADD . /code/

    This `Dockerfile` starts with a Python 2.7 image. The base Python image is
    modified by adding a new `code` directory. Into the directory, is installs
    the Python requirements defined in the `requirements.txt` file.   

4. Save and close the `Dockerfile`.

5. Create a `requirements.txt` in your project directory.

    This file is used by the `RUN pip install -r requirements.txt` command in your `Dockerfile`.

6. List the required software in the file.

        Django
        psycopg2

7. Save and close the `requirements.txt` file.

8. Create a file called `docker-compose.yml` in your project directory.

    The `docker-compose.yml` file describes the services that make your app. In
    this example those services are a web server and database.  The compose file
    also describes which Docker images these services use, how they link
    together, any volumes they might need mounted inside the containers.
    Finally, the `docker-compose.yml` file describes which ports these services
    expose. See the [`docker-compose.yml` reference](yml.md) for more
    information on how this file works.

9. Add the following configuration to the file.

      db:
        image: postgres
      web:
        build: .
        command: python manage.py runserver 0.0.0.0:8000
        volumes:
          - .:/code
        ports:
          - "8000:8000"
        links:
          - db

    This file defines two services. The `db` service and the `web` service.

10. Save and close the `docker-compose.yml` file.

## Create a Django project

In this step, you create a Django project by building the image from the build context you provided in the previous procedure.

1. Change to the root of your project directory.

2. Run the application using the `docker-compose` command.

        $ docker-compose run web django-admin.py startproject composeexample .

    This instructions Compose to the run `django-admin.py startproject
    composeexample` in a container. Before it can do this, Compose needs to know
    what to build and it looks in the current directory for instructions. The
    `docker-compose.yml` defines two services, the `web` service defines a `build`
    context as the current (.) directory. Finding the Dockerfile, Compose uses
    that to build the `web` service.

    Once the `web` service image is built, Compose runs it and executes the
    `django-admin.py startproject` command in the container. This command
    writes a set of files representing a Django app into your project
    directory:

3. After the `docker-compose` command completes, list the contents of your project.

        $ ls
        Dockerfile       docker-compose.yml          composeexample       manage.py        requirements.txt

## Connect the database

In this section, you set up the database connection for Django.

1. In your project dirctory, edit the `composeexample/settings.py` file.

2. Replace the `DATABASES = ...` with the following:

        DATABASES = {
            'default': {
                'ENGINE': 'django.db.backends.postgresql_psycopg2',
                'NAME': 'postgres',
                'USER': 'postgres',
                'HOST': 'db',
                'PORT': 5432,
            }
        }

    These settings are determined by the
    [postgres](https://registry.hub.docker.com/_/postgres/) Docker image
    specified in the Dockerfile.

3. Save and close the file.

4. Enter the `docker-compose up` command.

        $ docker-compose up
        Recreating myapp_db_1...
        Recreating myapp_web_1...
        Attaching to myapp_db_1, myapp_web_1
        myapp_db_1 |
        myapp_db_1 | PostgreSQL stand-alone backend 9.1.11
        myapp_db_1 | 2014-01-27 12:17:03 UTC LOG:  database system is ready to accept connections
        myapp_db_1 | 2014-01-27 12:17:03 UTC LOG:  autovacuum launcher started
        myapp_web_1 | Validating models...
        myapp_web_1 |
        myapp_web_1 | 0 errors found
        myapp_web_1 | January 27, 2014 - 12:12:40
        myapp_web_1 | Django version 1.6.1, using settings 'composeexample.settings'
        myapp_web_1 | Starting development server at http://0.0.0.0:8000/
        myapp_web_1 | Quit the server with CONTROL-C.

    At this point, your Django app should be running at port `8000` on your
    Docker host. If you are using a Docker Machine VM, you can use the
    `docker-machine ip MACHINE_NAME` to get the IP address.

5. Try running some Compose management commands with Docker.

    For example, open another terminal, change to your project directory and try the following commands.

        $ docker-compose up
        $ docker-compose run web python manage.py syncdb

## More Compose documentation

- [User guide](/)
- [Installing Compose](install.md)
- [Get started with Django](django.md)
- [Get started with Rails](rails.md)
- [Get started with Wordpress](wordpress.md)
- [Command line reference](/reference)
- [Yaml file reference](yml.md)
- [Compose environment variables](env.md)
- [Compose command line completion](completion.md)



--->

**Comment by charleswhchan** [comment]
@moxiegirl: Sorry for the delay, here's my feedback https://gist.github.com/charleswhchan/df3fbaa3ad6d6b3cd3d3. 

(By the way: I think you should integrate these changes and take the credit. However, I would be happy to update this PR as well -- if you prefer. :smiley:)

>        FROM python:2.7
>        ENV PYTHONUNBUFFERED 1
>        RUN mkdir /code
>        WORKDIR /code
>        ADD requirements.txt /code/
>        RUN pip install -r requirements.txt
>        ADD . /code/

Actually, I think we can simplify the 2x `ADD` commands. I tried the following, and the example still works. 
What do you think?
>        FROM python:2.7
>        ENV PYTHONUNBUFFERED 1
>        RUN mkdir /code
>        WORKDIR /code
>        ADD . /code/
>        RUN pip install -r requirements.txt

----

Step 3:
>    This `Dockerfile` starts with a Python 2.7 image. The base Python image is
>    modified by adding a new `code` directory. Into the directory, is installs
>    the Python requirements defined in the `requirements.txt` file. 

How about the following? 
>    This `Dockerfile` starts with a Python 2.7 base image. The base image is
>    modified by adding a new `code` directory. The base image is further modified
>    by installing the Python requirements defined in the `requirements.txt` file. 
    
----

> 6. List the required software in the file.

6. Add the required software into the file.

----

> In this step, you create a Django project by building the image from the build context you provided in the previous procedure.

In this step, you create a Django started project by building the image from the build context defined in the previous procedure.

----

> 2. Run the application using the docker-compose command.

2. Create the Django project using the docker-compose command.

----

>    This instructions Compose to the run `django-admin.py startproject
>    composeexample` in a container. Before it can do this, Compose needs to know
>    what to build and it looks in the current directory for instructions. The
>    `docker-compose.yml` defines two services, the `web` service defines a `build`
>    context as the current (.) directory. Finding the Dockerfile, Compose uses
>    that to build the `web` service.

>    Once the `web` service image is built, Compose runs it and executes the
>    `django-admin.py startproject` command in the container. This command
>    writes a set of files representing a Django app into your project
>    directory:

    This command instructs Compose to the run `django-admin.py startproject
    composeexample` in the `web` container. Before it can do this, Compose needs to know
    what to build and by default it looks for `docker-compose.yml` in the current directory for instructions. The
    `docker-compose.yml` defines two services, the `web` service and `db` service. 
    Since `web` service defines a `build` context as the current (.) directory, 
    Compose will also look for `Dockerfile` in the current directory, and uses that to build the `web` service.

    Once the `web` service image is built, Compose runs it and executes the
    `django-admin.py startproject` command in the container. This command
    instructs Django to create a set of files and directories representing a Django project.
    
----

>        $ docker-compose up
>        Recreating myapp_db_1...
>        Recreating myapp_web_1...
>        Attaching to myapp_db_1, myapp_web_1
>        myapp_db_1 |
>        myapp_db_1 | PostgreSQL stand-alone backend 9.1.11
>        myapp_db_1 | 2014-01-27 12:17:03 UTC LOG:  database system is ready to accept connections
>        myapp_db_1 | 2014-01-27 12:17:03 UTC LOG:  autovacuum launcher started
>        myapp_web_1 | Validating models...
>        myapp_web_1 |
>        myapp_web_1 | 0 errors found
>        myapp_web_1 | January 27, 2014 - 12:12:40
>        myapp_web_1 | Django version 1.6.1, using settings 'composeexample.settings'
>        myapp_web_1 | Starting development server at http://0.0.0.0:8000/
>        myapp_web_1 | Quit the server with CONTROL-C.

Updated messages:
>        docker-compose up
>        Starting composepractice_db_1...
>        Starting composepractice_web_1...
>        Attaching to composepractice_db_1, composepractice_web_1
>        ...
>        db_1  | PostgreSQL init process complete; ready for start up.
>        ...
>        db_1  | LOG:  database system is ready to accept connections
>        db_1  | LOG:  autovacuum launcher started
>        ...
>        web_1 | Django version 1.8.4, using settings 'composeexample.settings'
>        web_1 | Starting development server at http://0.0.0.0:8000/
>        web_1 | Quit the server with CONTROL-C.    
    
----

> 5. Try running some Compose management commands with Docker.

I think this step is beyond the scope of the quickstart guide. Remove perhaps?

----

Links:

> Get started with Django

Seems odd to page to link back to itself. Was this autogenerated? Should it be removed?

> Yaml file reference

Nit: YAML (upper case according to http://yaml.org/ and https://en.wikipedia.org/wiki/YAML)

**Comment by moxiegirl** [comment]
@charleswhchan Thanks! I'm perfectly happy for you to incorporate.  That's the benefit of working together, between the two of us we can arrive at something both technically correct and presented well.

I agree with all of your changes...except the one regarding the two `ADD` instructions... liked the original because it drew better attention to the inclusion of the `requirements.txt` file.  The original is not technically as elegant an execution, true, but better as a teaching example if that makes sense.

**Comment by charleswhchan** [comment]
@moxiegirl Incorporated all the structure changes from the gist, and the additional nits above -- **except** the proposal to simplified `ADD` statements.

Thanks again for the guidance! :+1:

File: docs/django.md [hunk 3]
```diff
@@ -10,126 +10,176 @@ weight=4
...
+
+In this step, you create a Django started project by building the image from the build context defined in the previous procedure.
+
+1. Change to the root of your project directory.
+
+2. Create the Django project using the `docker-compose` command.
+
+        $ docker-compose run web django-admin.py startproject composeexample .
+
+    This instructions Compose to the run `django-admin.py startproject
```
Comment (moxiegirl) [inline docs/django.md]: his instructions Compose > This instructs Compose

**Comment by moxiegirl** [comment]
@charleswhchan I see one small nit probably introduced by my comments. Otherwise LGTM

File: docs/django.md [hunk 4]
```diff
@@ -10,126 +10,176 @@ weight=4
...
-    ADD requirements.txt /code/
-    RUN pip install -r requirements.txt
-    ADD . /code/
+1. Create an empty project directory.
 
-This Dockerfile will define an image that is used to build a container that
-includes your application and has Python installed alongside all of your Python
-dependencies. For more information on how to write Dockerfiles, see the
-[Docker user guide](https://docs.docker.com/userguide/dockerimages/#building-an-image-from-a-dockerfile) and the [Dockerfile reference](http://docs.docker.com/reference/builder/).
+    You can name the directory something easy for your to remember. This directory is the context for your application image. The directory should only contain resources to build that image.
```
Comment (aanand) [inline docs/django.md]: Typo: your -> you

File: docs/django.md [hunk 5]
```diff
@@ -10,126 +10,176 @@ weight=4
...
+          build: .
+          command: python manage.py runserver 0.0.0.0:8000
+          volumes:
+            - .:/code
+          ports:
+            - "8000:8000"
+          links:
+            - db
+
+    This file defines two services. The `db` service and the `web` service.
```
Comment (aanand) [inline docs/django.md]: Wording this as two sentences is strange. I would have expected:

    This file defines two services: the `db` service and the `web` service.

File: docs/django.md (hunk 3 above)
Comment (aanand) [inline docs/django.md]: Also, "to the run" -> "to run"

**Comment by charleswhchan** [comment]
@aanand, @moxiegirl: Once again, thanks! 

I've addressed the latest round of issues (4). 
Have another read if you don't mind?

File: docs/django.md [hunk 6]
```diff
@@ -10,126 +10,176 @@ weight=4
...
+2. Create the Django project using the `docker-compose` command.
+
+        $ docker-compose run web django-admin.py startproject composeexample .
+
+    This instructions Compose to the run `django-admin.py startproject
+    composeexample` in a container. Before it can do this, Compose needs to know
+    what to build and it looks in the current directory for instructions. The
+    `docker-compose.yml` defines two services, the `web` service defines a `build`
+    context as the current (.) directory. Finding the Dockerfile, Compose uses
+    that to build the `web` service.
```
Comment (aanand) [inline docs/django.md]: This bit reads strangely to me. How about:

> This instructs Compose to run `django-admin.py startproject composeeexample` in a container, using the `web` service's image and configuration. Because the image doesn't exist yet, Compose builds it from the current directory, as specified by the `build: .` line in `docker-compose.yml`.

File: docs/django.md [hunk 7]
```diff
@@ -10,126 +10,176 @@ weight=4
...
+    the current directory for instructions. The `docker-compose.yml` defines two
+    services, the `web` service and `db` service. Since `web` service defines a
+    `build` context as the current (.) directory, Compose will also look for
+    `Dockerfile` in the current directory, and uses that to build the `web`
+    service.
+
+    Once the `web` service image is built, Compose runs it and executes the
+    `django-admin.py startproject` command in the container. This command
+    instructs Django to create a set of files and directories representing a
+    Django project.
```
Comment (aanand) [inline docs/django.md]: These paragraphs appear to be duplicated.

File: docs/django.md [hunk 8]
```diff
@@ -10,126 +10,176 @@ weight=4
...
+                'USER': 'postgres',
+                'HOST': 'db',
+                'PORT': 5432,
+            }
+        }
 
-Then, run `docker-compose up`:
+    These settings are determined by the
+    [postgres](https://registry.hub.docker.com/_/postgres/) Docker image
+    specified in the Dockerfile.
```
Comment (aanand) [inline docs/django.md]: `docker-compose.yml`, not `Dockerfile`.

File: docs/django.md [hunk 9]
```diff
@@ -10,126 +10,176 @@ weight=4
...
-    myapp_web_1 |
-    myapp_web_1 | 0 errors found
-    myapp_web_1 | January 27, 2014 - 12:12:40
-    myapp_web_1 | Django version 1.6.1, using settings 'composeexample.settings'
-    myapp_web_1 | Starting development server at http://0.0.0.0:8000/
-    myapp_web_1 | Quit the server with CONTROL-C.
+3. Save and close the file.
 
-Your Django app should nw be running at port 8000 on your Docker daemon. If you are using a Docker Machine VM, you can use the `docker-machine ip MACHINE_NAME` to get the IP address.
+4. Enter the `docker-compose up` command.
```
Comment (aanand) [inline docs/django.md]: Would "Run" be better than "Enter"?

File: docs/django.md [hunk 10]
```diff
@@ -10,126 +10,176 @@ weight=4
...
 
-You can also run management commands with Docker. To set up your database, for
-example, run `docker-compose up` and in another terminal run:
+        docker-compose up
```
Comment (aanand) [inline docs/django.md]: Should be a `$` before `docker-compose up`

File: docs/django.md (hunk 6 above)
Comment (charleswhchan) [inline docs/django.md]: Sure. I had hope to integrate `Dockfile` into the statements, but I think it is probably more details than necessary.

File: docs/django.md (hunk 7 above)
Comment (charleswhchan) [inline docs/django.md]: Good catch. Forgot to delete the old paragraphs after copy-n-paste. :+1:

File: docs/django.md (hunk 10 above)
Comment (charleswhchan) [inline docs/django.md]: Good catch.

File: docs/django.md (hunk 9 above)
Comment (charleswhchan) [inline docs/django.md]: Sure.

**Comment by charleswhchan** [comment]
PR updated with the latest round of fixes. :relaxed:

**Comment by moxiegirl** [comment]
Hi Charles, you have my approval. Check with @aanand for the final and merge.

**Comment by mnowster** [comment]
Hi @charleswhchan can you do a rebase please? Then I can give it a final review.

**Comment by charleswhchan** [comment]
Hi @mnowster, rebase as requested. 
Thanks in advance for the review, much appreciated! :relaxed:

**Comment by mnowster** [comment]
LGTM. Thanks @charleswhchan for the contribution :+1:

**Comment by charleswhchan** [comment]
Thanks everyone!
